[Curl] Represent IPv6 addresses according to RFC 5952 addressing architecture

mokjasmine opened the pull request:
#### 8bec657f93e0bcb721dd1a5f107e867174091976
<pre>
[Curl] Represent IPv6 addresses according to RFC 5952 addressing architecture
<a href="https://bugs.webkit.org/show_bug.cgi?id=244520">https://bugs.webkit.org/show_bug.cgi?id=244520</a>

Reviewed by Alex Christensen.

Conform to RFC 5952, <a href="https://www.rfc-editor.org/rfc/rfc4291">https://www.rfc-editor.org/rfc/rfc4291</a>, when displaying certificate
information.

* Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
(OpenSSL::getSubjectAltName):
(OpenSSL::getIPv6):
* Source/WebCore/platform/network/curl/OpenSSLHelper.h:
* Tools/TestWebKitAPI/PlatformPlayStation.cmake:
* Tools/TestWebKitAPI/PlatformWin.cmake:
* Tools/TestWebKitAPI/Tests/WebCore/curl/OpenSSLHelperTests.cpp: Added.
(TestWebKitAPI::Curl::OpenSSLHelperTests::OpenSSLHelperTests):
(TestWebKitAPI::Curl::TEST):

Canonical link: <a href="https://commits.webkit.org/255147@main">https://commits.webkit.org/255147@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ded5fbebce727e7e147b34991d218842d93e043

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91506 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/639 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22149 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/101233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/161269 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95511 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/662 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83849 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/97585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97164 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/27374 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/82343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82001 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/70412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35635 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33417 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/17122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3575 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37225 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39140 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36239 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
```diff
@@ -303,17 +303,79 @@ static void getSubjectAltName(const X509* x509, Vector<String>& dnsNames, Vector
 
+String getIPv6(String ip) 
+{
+    bool compressCurrSection = true;
```
We try to not use such abbreviations in our variable names.
compressCurrentSection

mokjasmine replied:
Addressed in https://github.com/WebKit/WebKit/pull/3791/commits/f8bae992ac33b2a826fce129d8884429633ede96

achristensen07 commented on Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
```diff
@@ -303,17 +303,79 @@ static void getSubjectAltName(const X509* x509, Vector<String>& dnsNames, Vector
+    String sections[8];
+    size_t maxZeros = 0;
+    size_t unset = -1;
+    size_t rstart = unset, rend = unset, start = unset, end = unset;
```
We also usually do one variable declaration per line.
Perhaps you're looking for std::optional<size_t>

mokjasmine replied:
Addressed in https://github.com/WebKit/WebKit/pull/3791/commits/f8bae992ac33b2a826fce129d8884429633ede96

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebCore/curl/OpenSSLHelperTests.cpp:
```diff
@@ -0,0 +1,87 @@
+    OpenSSLHelperTests() { }
+};
+
+TEST(OpenSSLHelper, IPv6AddressCompression)
```
I suggest that you add all the IPv6 test strings from the existing API test WTF_URLParser.Basic
They don't each need a comment explaining what is going on.

mokjasmine replied:
I don't feel like the IPv6 related tests in `WTF_URLParser.Basic` test anything that the current tests aren't already covering.

achristensen07 replied:
Oh, you're right, those tests take as input something that might already be compressed, and yours always takes an uncompressed input.

Add a test like 1:0:0:1:0:0:1:0 to verify that the correct set of two 0's is compressed.

mokjasmine replied:
Addressed in https://github.com/WebKit/WebKit/pull/3791/commits/75284d26b864b8943726f00329e296f662dbcdf3

achristensen07 commented on Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
```diff
@@ -303,17 +303,79 @@ static void getSubjectAltName(const X509* x509, Vector<String>& dnsNames, Vector
     }
 }
 
+String getIPv6(String ip) 
```
I think the parameter can be a const String&
This name isn't very descriptive of what the function does.  canonicalizeIPv6Address?

mokjasmine replied:
Addressed in https://github.com/WebKit/WebKit/pull/3791/commits/f8bae992ac33b2a826fce129d8884429633ede96

achristensen07 commented on Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
```diff
@@ -303,17 +303,79 @@ static void getSubjectAltName(const X509* x509, Vector<String>& dnsNames, Vector
                     data += 2;
                 }
-                ipAddresses.append(ipAddress.toString());
+                String ipv6 = getIPv6(ipAddress.toString());
```
Rather than appending integers to a string just to be parsed again, why don't you pass a std::array<uint16_t, 8> as a parameter?

mokjasmine replied:
Addressed in https://github.com/WebKit/WebKit/pull/3791/commits/f8bae992ac33b2a826fce129d8884429633ede96

achristensen07 commented on Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
```diff
@@ -301,19 +301,76 @@ static void getSubjectAltName(const X509* x509, Vector<String>& dnsNames, Vector
     }
 }
 
+String canonicalizeIPv6Address(String *ipInput) 
```
This signature is too open ended.  It could point to any number of Strings, and they could contain anything.  I think it would be much better to pass a const std::array<uint16_t, 8>&

mokjasmine replied:
Addressed in https://github.com/WebKit/WebKit/pull/3791/commits/75284d26b864b8943726f00329e296f662dbcdf3

achristensen07 commented on Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
```diff
@@ -301,19 +301,74 @@ static void getSubjectAltName(const X509* x509, Vector<String>& dnsNames, Vector
     }
 }
 
+String canonicalizeIPv6Address(unsigned char *data)
```
This function signature still does not indicate the length.  I think it would be far superior to use a function signature like this:
String canonicalizeIPV6Address(Span<uint8_t, 16>)
Otherwise, this function should be robust against any length of data being its input.

mokjasmine replied:
Addressed in https://github.com/WebKit/WebKit/pull/3791/commits/d4f739bfcbb4681199f23013d7ac903f78aff943

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebCore/curl/OpenSSLHelperTests.cpp:
```diff
@@ -0,0 +1,91 @@
+#if USE(CURL)
+
+#include <WebCore/OpenSSLHelper.h>
+#include <wtf/text/StringBuilder.h>
```
I think this is unused.

achristensen07 commented on Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
```diff
@@ -301,19 +301,76 @@ static void getSubjectAltName(const X509* x509, Vector<String>& dnsNames, Vector
+    int dataIndex = 0;
+
+    for (int i = 0; i < 8; i++) {
+        oldSection = ""_s;
```
This line is always unused.

achristensen07 commented on Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
```diff
@@ -301,19 +301,76 @@ static void getSubjectAltName(const X509* x509, Vector<String>& dnsNames, Vector
+    StringBuilder newSection;
+    size_t maxZeros = 0;
+    size_t unset = -1;
+    size_t rstart = unset;
```
Can these be given better names? "rstart" and "rend" doesn't really describe what they're used for.
Also, I think it would be better if they were std::optional<size_t> instead of using -1

achristensen07 commented on Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
```diff
@@ -301,19 +301,76 @@ static void getSubjectAltName(const X509* x509, Vector<String>& dnsNames, Vector
+        compressCurrentSection = true;
+        newSection = StringBuilder();
+        for (int j = 0; j < oldSection.length(); j++) {
+            if (oldSection[j] != '0')
```
We shouldn't need to do string comparisons.  Can't we look at the Span instead?

achristensen07 commented on Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
```diff
@@ -301,19 +301,64 @@ static void getSubjectAltName(const X509* x509, Vector<String>& dnsNames, Vector
+            j = end.value();
+            continue;
+        }
+        oldSection = makeString(hex(data[2 * j] << 8 | data[2 * j + 1], 4, Lowercase));
```
This still does several unnecessary string allocations.  This could be an integer operation.
Much better than it was before, and since it's only used in a rare case in the cURL code, I think it's fine to add it.

mokjasmine committed:
[Curl] Represent IPv6 addresses according to RFC 5952 addressing architecture
https://bugs.webkit.org/show_bug.cgi?id=244520

Reviewed by Alex Christensen.

Conform to RFC 5952, https://www.rfc-editor.org/rfc/rfc4291, when displaying certificate
information.

* Source/WebCore/platform/network/curl/OpenSSLHelper.cpp:
(OpenSSL::getSubjectAltName):
(OpenSSL::getIPv6):
* Source/WebCore/platform/network/curl/OpenSSLHelper.h:
* Tools/TestWebKitAPI/PlatformPlayStation.cmake:
* Tools/TestWebKitAPI/PlatformWin.cmake:
* Tools/TestWebKitAPI/Tests/WebCore/curl/OpenSSLHelperTests.cpp: Added.
(TestWebKitAPI::Curl::OpenSSLHelperTests::OpenSSLHelperTests):
(TestWebKitAPI::Curl::TEST):

Canonical link: https://commits.webkit.org/255147@main
